### DISCUSSION TITLE
Serializer : Deserialized a json with Collection and no 'id' on items

### DISCUSSION THREAD

**Discussion by piedrot** [symfony/symfony#46669] [Q&A]
Hello everyone,

I would like to have some help on a problem I can't manage to resolve with the serializer :

Update a 2 levels relation entity with a json object that I receive and who doesn't have the id from the database, only ObjetID.

Entities relations : 
<pre>
Task
    TaskService (Collection)
        TaskServicePeriodPlan (Collection)
</pre>

Serializer configuration :
<pre>
$task = $taskRepository->findOneBy(['objectID' => 'FA163E21C4E71EDCABD41C53EFF776BF']);
$payload = file_get_contents('/srv/api/tests/payload_task.json');

$classMetadataFactory  = new ClassMetadataFactory( new AnnotationLoader( new AnnotationReader() ) );
$nameConverter         = new PascalCaseNameConverter();
$normalizer            = new ObjectNormalizer( $classMetadataFactory, $nameConverter, null, new PropertyInfoExtractor() );
$serializer            = new Serializer( [$normalizer, new SapDateTimeNormalizer(), new ArrayDenormalizer() ], [new JsonEncoder( new JsonEncode() )] );

$updatedTask = $serializer->deserialize($payload, Task::class, 'json', [
            AbstractNormalizer::GROUPS => ['task:sap:put'],
            AbstractNormalizer::OBJECT_TO_POPULATE => $task,
            AbstractObjectNormalizer::DEEP_OBJECT_TO_POPULATE => true,
            SapDateTimeNormalizer::FORMAT_KEY => "U"
        ]);
</pre>

Here is how the TaskService is declared in Task entity:
<pre>
    #[ApiSubresource]
    #[Groups(["project:read", "project:put", "task:read", "task:put",'task:sap:put'])]
    #[ORM\OneToMany( mappedBy: 'task', targetEntity: TaskService::class, cascade: ['persist'], fetch: 'EAGER', orphanRemoval: true )]
    private Collection $taskService;
</pre>

A simplify version of the JSON with 2 levels relation :
<pre>
{
    "ObjectID": "FA163E21C4E71EDCABD41C53EFF776BF",
    "ParentObjectID": "FA163E21C4E71EDCABD18B1F6B528B04",
    "TaskUUID": "FA163E21-C4E7-1EDC-ABD4-1C53EFF776BF",
    "TaskID": "CPSO44-16",
    "TaskName": "Modélisation D",
    "TaskService": [
        "ObjectID": "FA163E21C4E71EDCAC9E3C846B2A4110",
        "ParentObjectID": "FA163E21C4E71EDCABD41C53EFF776BF",
        "ProductID": "S200105",
        "TaskServicePeriodPlan": [
        {
          "ObjectID": "FA163E21C4E71EDCAEB419DC120C538F",
          "ParentObjectID": "FA163E21C4E71EDCAC9E3C846B2A4110",
          "PeriodStartDateTime": "/Date(1648425600000)/",
        },
        {
          "ObjectID": "FA163E21C4E71EDCAEB419DC120C538F",
          "ParentObjectID": "FA163E21C4E71EDCAC9E3C846B2A4110",
          "PeriodStartDateTime": "/Date(1648425600000)/",
        }],
    ],
}
</pre>

And here is a short dump of what i get from $updatedTask ( productID is supposed to be changed to S200105 ) :
<pre>
[App\Entity\Task {#2978 ▼
  -id: 3
  -taskID: "CPSO44-16"
  -objectID: "FA163E21C4E71EDCABD41C53EFF776BF"
  -taskName: "Modélisation D"
  -plannedDuration: "P78D"
  -isMain: false
  -createdAt: DateTimeImmutable @1655042387 {#2968 ▶}
  -updatedAt: DateTime @1655042387 {#2976 ▶}
  -project: Proxies\__CG__\App\Entity\Project {#3027 ▶}
  -manager: Proxies\__CG__\App\Entity\User {#3069 ▶}
  -taskService: Doctrine\ORM\PersistentCollection {#3035 ▼
    #collection: Doctrine\Common\Collections\ArrayCollection {#3034 ▼
      -elements: array:4 [▼
        0 => App\Entity\TaskService {#3012 ▼
          -id: 1
          -objectID: "FA163E21C4E71EDCAC9E3C846B2A4110"
          -productID: "S200101"
          -task: App\Entity\Task {#2978}
          -taskServicePeriodPlan: Doctrine\ORM\PersistentCollection {#3076 ▶}
          -serviceProduct: Proxies\__CG__\App\Entity\ServiceProduct {#3095 ▶}
          -user: Proxies\__CG__\App\Entity\User {#3094 ▶}
</pre>

For what I understood, to achieve relation deserialization, it is necessary to activate OBJECT_TO_POPULATE and DEEP_OBJECT_TO_POPULATE which I did.

But for now, the serializer is only able to update the properties of the task entity but is not able to update the TaskService Collection neither the TaskServicePeriodPlan Collection.

One of the things I don't get is the fact that the serializer is not creating new TaskService. As I guess he's not able to find a match with an id I would have guess that he would create new TaskService entity, am I wrong ?
And if I add a new TaskService in the payload, none is added to the TaskService Collection either.

So the first thing I would like to understand is why there is no new TaskService entity added to the Collection.

And then find a way to tell the serializer to find a match with a Collection based on ObjectID in the place of usual id to update the database task entity.
I'm not even sure it's doable as the documentation is not very clear to me : [Desierializer](https://symfony.com/doc/current/components/serializer.html#deserializing-in-an-existing-object)

I hope I'm clear enough, if not, don't hesitate to tell me.

The project is based on api-platform 2.6 but I'm using the classic Serializer

Thank You !

**Comment by lyrixx** [comment]
If you are not using API Platform, you'll have to create your own denormalizer, to search the data in the DB, or to created it when needed.

With API Platform, it's automatic.

---

If you think everything is OK, you have to ensure

* the property can be writeable (public property, or public setter)
* if it's a setter, the property name matches the method name (I made this type way too many times 😂 ): `foobar` => `setFoobar` (and not `setFobar` for instance) 
* the serialization group are the same all way long, from the very first object to the last one
* There is a way for the serializer to know how to denormalize the object (typed property; PHPDoc for collection `@var TaskService[]; type hint on the getter, setter, added, remover, etc; depending of the metadata extraction, you have to install some dependencies like php documentor, phpstan, etc)
* If you construct the serializer manually (not using the one in the full stack framework), you shoud check if all metadata extractor are filled.

If you cannot find the error, you can create a reproducer. Someone **may** help you :)

**Reply by piedrot** [reply]
Hi lyrixx, thank for your quick reply.

If I'm not using API-Platform on this part it's because it's related to a thrid-party system. I have to retrieve data from this remote system and store them in database. So I can't rely on API-Platform workflow here or if there is a way I don't get it...

I tought about the denormalizer as it is explain in the doc and on some stackoverflow post but I wasn't convince about this idea and not sure about how to do it. But I gave it a try and here is what came out of it :

<pre>
// CustomNormalizer.php
class TaskServiceNormalizer implements DenormalizerInterface, CacheableSupportsMethodInterface
{

    const entities = [
        Task::class,
        TaskService::class,
        TaskServicePeriodPlan::class,
    ];
    
    private EntityManagerInterface $em;
    private SapDateTimeNormalizer $sapDateTimeNormalizer;

    public function __construct( EntityManagerInterface $em, SapDateTimeNormalizer $sapDateTimeNormalizer )
    {
        $this->sapDateTimeNormalizer = $sapDateTimeNormalizer;
        $this->em                    = $em;
    }

    /**
     * @inheritdoc
     */
    public function supportsDenormalization( $object, $type, $format = null ): bool
    {
        return in_array( $type, self::entities );
    }

    /**
     * @inheritDoc
     */
    public function denormalize( $data, $class, $format = null, array $context = [] ): object
    {
        $dbEntity = $this->em->getRepository( $class )->findOneBy( ['objectID' => $data[ 'ObjectID' ]] );

        return $dbEntity
            ? $this->updateEntity( $data, $class, $context, $dbEntity )
            : $this->createEntity( $data, $class, $context );
    }

    private function updateEntity( $data, $class, $context, $dbEntity ): object
    {
        $reflectionClass = new \ReflectionClass( $class );
        foreach ( $reflectionClass->getProperties() as $property ) {
            $reflectionPropName = ucfirst( $property->getName() );

            if ( !isset( $data[ $reflectionPropName ] ) ) {
                continue;
            }

            // We only update the fields that are related to the groups in context
            if ( count( $property->getAttributes( 'Symfony\Component\Serializer\Annotation\Groups' ) ) ) {
                $reflectionPropGroups = $property->getAttributes( 'Symfony\Component\Serializer\Annotation\Groups' )[ 0 ]->getArguments()[ 0 ];
                $sameGroup            = array_intersect( $reflectionPropGroups, $context[ 'groups' ] );
                if ( count( $sameGroup ) ) {
                    $dbEntity = $this->hydrateEntity($data, $property->getType()->getName(), $context, $dbEntity, $reflectionPropName);
                }
            }
        }

        return $dbEntity;
    }

    private function createEntity( $data, $class, $context ): object
    {
        $reflectionClass = new \ReflectionClass( $class );
        $newEntity       = new $class();
        $propsName       = array_keys( $data );

        foreach ( $reflectionClass->getProperties() as $property ) {
            $reflectionPropName = ucfirst( $property->getName() );
            if ( in_array( $reflectionPropName, $propsName ) ) {
                $newEntity = $this->hydrateEntity($data, $property->getType()->getName(), $context, $newEntity, $reflectionPropName);
            }
        }

        return $newEntity;
    }

    private function hydrateEntity( $data, $type, $context, $entity, $reflectionPropName )
    {
        $propSetterName = ( is_array( $data[ $reflectionPropName ] ) ? 'add' : 'set' ) . $reflectionPropName;

        switch ( $type ) {
            case 'string':
                $entity->$propSetterName( (string)$data[ $reflectionPropName ] );
                break;
            case 'DateTimeInterface':
                $entity->$propSetterName( $this->sapDateTimeNormalizer->denormalize( $data[ $reflectionPropName ],
                                                                                       'DateTime', null,
                                                                                       $context ) );
                break;
            case 'DateTimeImmutable':
                $entity->$propSetterName( $this->sapDateTimeNormalizer->denormalize( $data[ $reflectionPropName ],
                                                                                       'DateTimeImmutable',
                                                                                       null, $context ) );
                break;
            case 'Doctrine\Common\Collections\Collection':
                if ( count( $data[ $reflectionPropName ] ) ) {
                    foreach ( $data[ $reflectionPropName ] as $dataEntity ) {
                        $entity->$propSetterName( $this->denormalize( $dataEntity,
                                                                      'App\Entity\\' . $reflectionPropName,
                                                                      null, $context ) );
                    }
                }
                break;
        }
        return $entity;
    }
</pre>

I'm still not sure it is the right way to use it and it seems I'm not using the Serializer as it should be, I don't know.
And I still need to test more to see if it's bullet proof cause right now it's just a draft...

I'm open to any recommendations even if I know it's not the purpose of this Discussions channel to debug my code ><.

Regards

**Reply by lyrixx** [reply]
>I'm still not sure it is the right way to use it and it seems I'm not using the Serializer as it should be, I don't know.

Yes and no :) But it's too hard to explain here. I deserves a training / coaching.

TL;DR: The denormalize idea is good. but IMHO you should not use the reflection and you should let the serializer to call itself your recursion

**Reply by piedrot** [reply]
Yeah I know I could use some training / coaching on this topic cause it's not so easy to master ^^

Anyway, I'll try to make some change on this denormalizer and see if I can make it work with serializer without doing the recursion myself.

Thank you for the tip !
